toolbox/HBTReader.py

```diff
@@ -1,279 +1,686 @@
+            HostHaloId (int): host halo id
+            isnap (int): (default=-1) snapshot number
+
+        Returns:
+            (numpy.ndarray): array of host halo ids of progenitors
+        """
+
+        try:
+            subhaloes = self.GetSubsOfHost(HostHaloId, isnap=isnap)['TrackId']
+            progenitors = np.unique([
+                self.GetSub(subhalo, isnap - 1)[0]['HostHaloId']
+                for subhalo in subhaloes
+                if self.GetSub(subhalo, isnap - 1)[0]['Rank'] == 0
+            ])
+        except:
+            progenitors = np.array([])
+
+        return progenitors
+
+    def GetCollapsedMassHistory(self, HostHaloId, isnap=-1, f=0.01):
+        """Calculates a CMH, starting at a FOF group.
+
+        CMH is a sum of masses of all progenitors over a threshold.
+
+        Arguments:
+            HostHaloId (int): starting point of the tree
+            isnap (int): (default = -1) initial snapshot
+            f (float): (default = 0.01) NFW :math:`f` parameter
+
+        Returns:
+            (numpy.ndarray): CMH of a host halo
+        """
+
+        m0 = self.GetHostHalo(HostHaloId, isnap)['M200Crit']
+        trackIds = self.GetSubsOfHost(HostHaloId, isnap)['TrackId']
+
+        logging.info(
+            "Starting halo %d of mass %.2f with %d tracks at snapshot %d" %
+            (HostHaloId, m0, len(trackIds), isnap))
+
+        hosts = []
+        for trackId in trackIds:
+            track = self.GetTrack(trackId, MaxSnap=isnap)
+            logging.debug("Track %d of halo %d@%d across %d snapshots" %
+                          (trackId, HostHaloId, isnap, len(track)))
+            hosts_of_track = zip(track['Snapshot'], track['HostHaloId'])
+            hosts.extend(hosts_of_track)
+
+        hosts = np.unique(
+            np.array(
+                hosts, dtype=np.dtype([('Snapshot', int), ('HaloId', int)])),
+            axis=0)
+        hosts = hosts[hosts['HaloId'] != 0]
+
+        masses = [
+            self.GetHostHalo(host['HaloId'], host['Snapshot'])['M200Crit']
+            for host in hosts
+        ]
+        hosts = append_fields(hosts, 'M200Crit', masses, usemask=False)
+        snaps = np.unique(hosts['Snapshot'])
+
+        logging.debug(
+            "Queried halo %d@%d with %d host halos across %d snapshots" %
+            (HostHaloId, isnap, len(hosts), len(snaps)))
+
+        cmh = np.array(
+            zip(np.full_like(snaps, HostHaloId), snaps, np.zeros_like(snaps)),
+            dtype=np.dtype([
+                ('HaloId', np.int32),
+                ('Snapshot', np.int32),
+                ('M200Crit', np.float32),
+            ]))
+        for i, _ in np.ndenumerate(cmh):
+            cmh[i]['M200Crit'] = np.sum(filter(lambda m: m > f * m0,\
+                hosts[hosts['Snapshot'] == cmh[i]['Snapshot']]['M200Crit']))
+
+        logging.info("Finished CMH for halo %d@%d" % (HostHaloId, isnap))
+
+        return cmh
+
+    def GetHostProfile(self, selection=None, isnap=-1):
```

Comment:
the profiles may not always be available. If yes, then it can also be accessed with selection in LoadHostHalos() or after GetHostHalo(). So it's best to keep this outside the general HBTReader.